model/infinispan/src/main/java/org/keycloak/crls/infinispan/InfinispanCrlStorageProviderFactory.java

```diff
@@ -0,0 +1,148 @@
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.keycloak.crls.infinispan;
+
+import org.infinispan.Cache;
+import org.infinispan.configuration.cache.ConfigurationBuilder;
+import org.infinispan.manager.DefaultCacheManager;
+import org.jboss.logging.Logger;
+import org.keycloak.Config;
+import org.keycloak.crls.CrlLoader;
+import org.keycloak.crls.CrlStorageProvider;
+import org.keycloak.crls.CrlStorageProviderFactory;
+import org.keycloak.models.KeycloakSession;
+import org.keycloak.models.KeycloakSessionFactory;
+import org.keycloak.provider.EnvironmentDependentProviderFactory;
+import org.keycloak.crls.CrlEntry;
+import org.keycloak.provider.ProviderConfigProperty;
+import org.keycloak.provider.ProviderConfigurationBuilder;
+
+import java.io.IOException;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Future;
+
+import static org.keycloak.connections.infinispan.InfinispanConnectionProvider.CRLS_CACHE_NAME;
+
+/**
+ * Crl Storage Provider Factory backed by Infinispan cache
+ *
+ * @author Joshua Smith
+ * @author Scott Tustison
+ */
+public class InfinispanCrlStorageProviderFactory implements CrlStorageProviderFactory, EnvironmentDependentProviderFactory {
+
+    private static final Logger log = Logger.getLogger(InfinispanCrlStorageProviderFactory.class);
+
+    public static final String PROVIDER_ID = "infinispan";
+
+    public enum CacheExpirationMode {
+        NEVER_EXPIRE,
+        NEXT_UPDATE,
+        MAX_CACHE_TIME
+    }
+
+    private volatile Cache<String, CrlEntry> crlCache;
+
+    private DefaultCacheManager cacheManager;
+
+    private final Map<String, Future<CrlEntry>> tasksInProgress = new ConcurrentHashMap<>();
+
+    private CacheExpirationMode cacheExpirationMode;
+
+    private long maxCacheTime;
+
+    @Override
+    public CrlStorageProvider create(KeycloakSession session) {
+        lazyInit(session);
+        return new InfinispanCrlStorageProvider(session, crlCache, tasksInProgress, new CrlLoader(session), cacheExpirationMode, maxCacheTime);
+    }
+
+    private void lazyInit(KeycloakSession session) {
+        if (crlCache == null) {
+            synchronized (this) {
+                if (crlCache == null) {
+                    cacheManager = new DefaultCacheManager();
+                    cacheManager.defineConfiguration(CRLS_CACHE_NAME, new ConfigurationBuilder().build());
```

Comment:
The addition of the new infinispan cache may need to be reviewed by @ahus1 or someone from the SRE team.

IMO it may be better to re-use the existing infinispan `CacheManager` instead of instantiating new CacheManager. In relation to this, it is probably better to integrate new cache inside `DefaultInfinispanConnectionProviderFactory` . It seems to me that it is quite similar to the "keys" cache . I suppose that new cache does not need to have expiration (As expiration is handled by the dedicated method in `InfinispanCrlStorageProvider.loadCrl`), however it may probably still have some eviction to make sure that cache does not accidentally fill whole memory. Maybe some new method similar to `getKeysCacheConfig()` can be added to `DefaultInfinispanConnectionProviderFactory` for handling this?

@ahus1 What do you think?